### PR TITLE
Add ARM/MOTOR_EMERGENCY_STOP Aux Switch

### PR DESCRIPTION
**Pull request by Hwurzburg**
a safer version of ARM/DISARM....which I have on every vehicle....and have accidentally hit once or twice in the air, generating frantic activity....

SITL tested in Copter and Plane...
High requests ARM and logs that Emergency Motor Stop is deactivated
Low activates and logs Emergency Motor Stop....in Plane you dont have a time limit to move back high and restart motors in case of accidental activation in the air...this position is a pre-arm fail in Copter and Plane
in Copter you have disarm_delay parameter/2 time to move back to High while in the air before a disarm happens (too short in my opinion, but thats current behavior)
MID de-activates motor stop, so as to check if disarm has occurred after landing/crashing but not request a re-arm
also checks that this switch and interlock in heli dont co-exist since it can call an Emergency Motor Stop, not allowed with an interlock switch in heli

### CODE REVIEW

**Comment by timtuxworth** [comment]
This seems interesting @Hwurzburg . Would I put it on my "throttle cut" switch on my Transmitter? I like that high is "request Arm" - with Yaapu telemetry it will advise me if this was successful.

**Comment by Hwurzburg** [comment]
put it on any switch...I dont know what you have setup for "throttle cut" now

File: libraries/RC_Channel/RC_Channel.cpp [hunk 1]
```diff
@@ -1250,6 +1251,34 @@ bool RC_Channel::do_aux_function(const aux_func_t ch_option, const AuxSwitchPos
...
+           case AuxSwitchPos::MIDDLE:
+               // nothing
+               break;
+           case AuxSwitchPos::LOW:
+               SRV_Channels::set_emergency_stop(true);
+               // log E-stop
+               AP_Logger *logger = AP_Logger::get_singleton();
+               if (logger && logger->logging_enabled()) {
+                   logger->Write_Event(LogEvent::MOTORS_EMERGENCY_STOPPED);
+               }
```
Comment (IamPete1) [inline libraries/RC_Channel/RC_Channel.cpp:1276]: Might be worth moving this logging down into servo channels now it is in two places.

File: libraries/RC_Channel/RC_Channel.cpp [hunk 2]
```diff
@@ -1250,6 +1251,34 @@ bool RC_Channel::do_aux_function(const aux_func_t ch_option, const AuxSwitchPos
         break;
     }
 
+    case AUX_FUNC::ARM_EMERGENCY_STOP: {
+        switch (ch_flag) {
+           case AuxSwitchPos::HIGH:
+              SRV_Channels::set_emergency_stop(false);
+              AP::arming().arm(AP_Arming::Method::AUXSWITCH, true);
+              if (hal.util->get_soft_armed()) {
```
Comment (IamPete1) [inline libraries/RC_Channel/RC_Channel.cpp:1259]: This always disables E stop so it should always log?

File: libraries/RC_Channel/RC_Channel.cpp [hunk 3]
```diff
@@ -1250,6 +1251,34 @@ bool RC_Channel::do_aux_function(const aux_func_t ch_option, const AuxSwitchPos
         break;
     }
 
+    case AUX_FUNC::ARM_EMERGENCY_STOP: {
+        switch (ch_flag) {
+           case AuxSwitchPos::HIGH:
+              SRV_Channels::set_emergency_stop(false);
+              AP::arming().arm(AP_Arming::Method::AUXSWITCH, true);
```
Comment (IamPete1) [inline libraries/RC_Channel/RC_Channel.cpp:1258]: I wounder what the behavior should be if arming fails. Should E-Stop be left off or re-engaged?

File: libraries/RC_Channel/RC_Channel.cpp (hunk 1 above)
Comment (Hwurzburg) [reply libraries/RC_Channel/RC_Channel.cpp:1276]: I can do that

File: libraries/RC_Channel/RC_Channel.cpp (hunk 2 above)
Comment (Hwurzburg) [reply libraries/RC_Channel/RC_Channel.cpp:1259]: just following E-Stop..logs on both true and false

File: libraries/RC_Channel/RC_Channel.cpp (hunk 3 above)
Comment (Hwurzburg) [reply libraries/RC_Channel/RC_Channel.cpp:1258]: if left re-engaged, then you cant arm from another method...thinks its best left disengaged if switch is high...its really the union of a arm switch and an inverted estop switch...

File: libraries/RC_Channel/RC_Channel.cpp (hunk 1 above)
Comment (Hwurzburg) [reply libraries/RC_Channel/RC_Channel.cpp]: done

File: libraries/SRV_Channel/SRV_Channel.cpp [hunk 4]
```diff
@@ -123,7 +125,24 @@ void SRV_Channel::calc_pwm(float output_scaled)
...
+      if (SRV_Channel::should_e_stop(get_function()) && !SRV_Channels::emergency_stop) { 
+         if (estop_active) {
+             AP_Logger *logger = AP_Logger::get_singleton();
+             if (logger && logger->logging_enabled()) {
+                 logger->Write_Event(LogEvent::MOTORS_EMERGENCY_STOP_CLEARED);
+             }
+             estop_active = false;
+         }   
+      } 
+    }
```
Comment (IamPete1) [inline libraries/SRV_Channel/SRV_Channel.cpp:145]: This should be in the `set_emergency_stop` function.

Comment (Hwurzburg) [reply libraries/SRV_Channel/SRV_Channel.cpp]: thanks for showing me how...

**Comment by MichelleRos** [comment]
Also useful is it looks like middle switch position will be neither e-stop nor arm. So if you land & e-stop, then you can move to middle position to check that it has disarmed (and perhaps then move it back to e-stop) before going to pick it up. (This is if you don't have a ground station or yaapu telemetry to announce it.)

File: libraries/SRV_Channel/SRV_Channel.cpp [hunk 5]
```diff
@@ -123,7 +123,7 @@ void SRV_Channel::calc_pwm(float output_scaled)
     if (SRV_Channel::should_e_stop(get_function()) && SRV_Channels::emergency_stop) {
         output_scaled = 0.0;
         force = true;
-    }
+    }
```
Comment (IamPete1) [inline libraries/SRV_Channel/SRV_Channel.cpp:126]: white space change left over

**Review by IamPete1 — COMMENTED** [review]
Looking good from a code point of view. Might be worth double checking that the E-stop enabled/disabled event logging still works as expected after the move.

**Comment by Hwurzburg** [comment]
yes...rechecked this and normal estop operation and logging

**Comment by Hwurzburg** [comment]
@MichelleRos middle position does nothing to change state....it could only disable estop, as would high, but not do an arm request...so if not disarmed,  motors would spin in copter and quadplane....
that might be nice...will add

**Comment by MichelleRos** [comment]
@Hwurzburg Ah you're right. The "middle" section of that case statement would need to also have SRV_Channels::set_emergency_stop(false) in order for that method to work.

**Comment by Hwurzburg** [comment]
@MichelleRos already added...thanks for the suggestion

**Review by rmackay9 — APPROVED** [review]

**Comment by rmackay9** [comment]
oops needs a rebase..

**Comment by peterbarker** [comment]
Merged - thanks!

**Pull request merged**